File: spec/javascript/grammar_spec.js
var jstest = require("jstest").Test

var Grammar = require("../../javascript/protocol/grammar")

jstest.describe("Grammar", function() { with(this) {
  describe("CHANNEL_NAME", function() { with(this) {
    it("matches valid channel names", function() { with(this) {
      assertMatch( Grammar.CHANNEL_NAME, "/fo_o/$@()bar" )
    }})

    it("does not match channel patterns", function() { with(this) {
      assertNoMatch( Grammar.CHANNEL_NAME, "/foo/**" )
    }})

    it("does not match invalid channel names", function() { with(this) {
      assertNoMatch( Grammar.CHANNEL_NAME, "foo/$@()bar" )
      assertNoMatch( Grammar.CHANNEL_NAME, "/foo/$@()bar/" )
      assertNoMatch( Grammar.CHANNEL_NAME, "/fo o/$@()bar" )
    }})
  }})

  describe("CHANNEL_PATTERN", function() { with(this) {
    it("does not match channel names", function() { with(this) {
      assertNoMatch( Grammar.CHANNEL_PATTERN, "/fo_o/$@()bar" )
    }})

    it("matches valid channel patterns", function() { with(this) {
      assertMatch( Grammar.CHANNEL_PATTERN, "/foo/**" )
      assertMatch( Grammar.CHANNEL_PATTERN, "/foo/*" )
    }})

    it("does not match invalid channel patterns", function() { with(this) {
      assertNoMatch( Grammar.CHANNEL_PATTERN, "/foo/**/*" )
    }})
  }})

  describe("ERROR", function() { with(this) {
    it("matches an error with an argument", function() { with(this) {
      assertMatch( Grammar.ERROR, "402:xj3sjdsjdsjad:Unknown Client ID" )
    }})

    it("matches an error with many arguments", function() { with(this) {
      assertMatch( Grammar.ERROR, "403:xj3sjdsjdsjad,/foo/bar:Subscription denied" )
    }})

    it("matches an error with no arguments", function() { with(this) {
      assertMatch( Grammar.ERROR, "402::Unknown Client ID" )
    }})

    it("does not match an error with no code", function() { with(this) {
      assertNoMatch( Grammar.ERROR, ":xj3sjdsjdsjad:Unknown Client ID" )
    }})

    it("does not match an error with an invalid code", function() { with(this) {
      assertNoMatch( Grammar.ERROR, "40:xj3sjdsjdsjad:Unknown Client ID" )
    }})
  }})

  describe("VERSION", function() { with(this) {
    it("matches a version number", function() { with(this) {
      assertMatch( Grammar.VERSION, "9" )
      assertMatch( Grammar.VERSION, "9.0.a-delta1" )
    }})

    it("does not match invalid version numbers", function() { with(this) {
      assertNoMatch( Grammar.VERSION, "9.0.a-delta1." )
      assertNoMatch( Grammar.VERSION, "" )
    }})
  }})
}})

File: spec/ruby/grammar_spec.rb
require "spec_helper"

describe Faye::Grammar do
  describe :CHANNEL_NAME do
    it "matches valid channel names" do
      Faye::Grammar::CHANNEL_NAME.should =~ "/fo_o/$@()bar"
    end

    it "does not match channel patterns" do
      Faye::Grammar::CHANNEL_NAME.should_not =~ "/foo/**"
    end

    it "does not match invalid channel names" do
      Faye::Grammar::CHANNEL_NAME.should_not =~ "foo/$@()bar"
      Faye::Grammar::CHANNEL_NAME.should_not =~ "/foo/$@()bar/"
      Faye::Grammar::CHANNEL_NAME.should_not =~ "/fo o/$@()bar"
    end
  end

  describe :CHANNEL_PATTERN do
    it "does not match channel names" do
      Faye::Grammar::CHANNEL_PATTERN.should_not =~ "/fo_o/$@()bar"
    end

    it "matches valid channel patterns" do
      Faye::Grammar::CHANNEL_PATTERN.should =~ "/foo/**"
      Faye::Grammar::CHANNEL_PATTERN.should =~ "/foo/*"
    end

    it "does not match invalid channel patterns" do
      Faye::Grammar::CHANNEL_PATTERN.should_not =~ "/foo/**/*"
    end
  end

  describe :ERROR do
    it "matches an error with an argument" do
      Faye::Grammar::ERROR.should =~ "402:xj3sjdsjdsjad:Unknown Client ID"
    end

    it "matches an error with many arguments" do
      Faye::Grammar::ERROR.should =~ "403:xj3sjdsjdsjad,/foo/bar:Subscription denied"
    end

    it "matches an error with no arguments" do
      Faye::Grammar::ERROR.should =~ "402::Unknown Client ID"
    end

    it "does not match an error with no code" do
      Faye::Grammar::ERROR.should_not =~ ":xj3sjdsjdsjad:Unknown Client ID"
    end

    it "does not match an error with an invalid code" do
      Faye::Grammar::ERROR.should_not =~ "40:xj3sjdsjdsjad:Unknown Client ID"
    end
  end

  describe :VERSION do
    it "matches a version number" do
      Faye::Grammar::VERSION.should =~ "9"
      Faye::Grammar::VERSION.should =~ "9.0.a-delta1"
    end

    it "does not match invalid version numbers" do
      Faye::Grammar::VERSION.should_not =~ "9.0.a-delta1."
      Faye::Grammar::VERSION.should_not =~ ""
    end
  end
end
